File: grimdb/sdk/ruby/lib/grimlocker/entries.rb
# frozen_string_literal: true
module Grimlocker
  Entry = Struct.new(:id, :title, :category, :fields, :created_at, :updated_at) do
    def self.from_hash(h)
      new(h['id']||'', h['title']||'', h['category']||'',
          h['fields']||{}, h['created_at']||0, h['updated_at']||0)
    end
    def field(key) = fields[key.to_s] || ''
  end

  PasswordEntry = Struct.new(:id, :title, :username, :password, :url, :notes) do
    def self.from_entry(e)
      new(e.id, e.title, e.field('username'), e.field('password'), e.field('url'), e.field('notes'))
    end
    def to_fields = { 'username' => username, 'password' => password, 'url' => url, 'notes' => notes }
  end

  SshKeyEntry = Struct.new(:id, :title, :public_key, :private_key, :username, :passphrase, :comment) do
    def self.from_entry(e)
      new(e.id, e.title, e.field('public_key'), e.field('private_key'),
          e.field('username'), e.field('passphrase'), e.field('comment'))
    end
    def to_fields = { 'public_key' => public_key, 'private_key' => private_key,
                      'username' => username, 'passphrase' => passphrase, 'comment' => comment }
  end

  CertificateEntry = Struct.new(:id, :title, :domain, :certificate, :private_key) do
    def self.from_entry(e)
      new(e.id, e.title, e.field('domain'), e.field('certificate'), e.field('private_key'))
    end
    def to_fields = { 'domain' => domain, 'certificate' => certificate, 'private_key' => private_key }
  end
end

File: grimdb/sdk/python/grimlocker/entries.py
"""Typed domain objects for vault entries."""

from dataclasses import dataclass, field
from typing import Dict


@dataclass
class Entry:
    id: str = ""
    category: str = ""
    title: str = ""
    fields: Dict[str, str] = field(default_factory=dict)
    created_at: int = 0
    updated_at: int = 0

    @classmethod
    def from_dict(cls, d: dict) -> "Entry":
        return cls(
            id=d.get("id", ""),
            category=d.get("category", ""),
            title=d.get("title", ""),
            fields=d.get("fields") or {},
            created_at=d.get("created_at", 0),
            updated_at=d.get("updated_at", 0),
        )


@dataclass
class PasswordEntry:
    title: str
    username: str = ""
    password: str = ""
    url: str = ""
    notes: str = ""
    id: str = ""

    @classmethod
    def from_entry(cls, e: Entry) -> "PasswordEntry":
        return cls(
            id=e.id,
            title=e.title,
            username=e.fields.get("username", ""),
            password=e.fields.get("password", ""),
            url=e.fields.get("url", ""),
            notes=e.fields.get("notes", ""),
        )

    def to_fields(self) -> Dict[str, str]:
        return {
            "username": self.username,
            "password": self.password,
            "url": self.url,
            "notes": self.notes,
        }


@dataclass
class SSHKeyEntry:
    title: str
    public_key: str = ""
    comment: str = ""
    algorithm: str = ""
    id: str = ""

    @classmethod
    def from_entry(cls, e: Entry) -> "SSHKeyEntry":
        return cls(
            id=e.id,
            title=e.title,
            public_key=e.fields.get("public_key", ""),
            comment=e.fields.get("comment", ""),
            algorithm=e.fields.get("algorithm", ""),
        )

    def to_fields(self) -> Dict[str, str]:
        return {
            "public_key": self.public_key,
            "comment": self.comment,
            "algorithm": self.algorithm,
        }


@dataclass
class CertificateEntry:
    title: str
    domain: str = ""
    certificate: str = ""
    private_key: str = ""
    id: str = ""

    @classmethod
    def from_entry(cls, e: Entry) -> "CertificateEntry":
        return cls(
            id=e.id,
            title=e.title,
            domain=e.fields.get("domain", ""),
            certificate=e.fields.get("certificate", ""),
            private_key=e.fields.get("private_key", ""),
        )

    def to_fields(self) -> Dict[str, str]:
        return {
            "domain": self.domain,
            "certificate": self.certificate,
            "private_key": self.private_key,
        }
